### PR TITLE
Channelz Part 10: More Channel Tracing

### PR DESCRIPTION
**Pull request by ncteisen**
Adds tracing for connectivity state change.

Tracking: #15340 

Example output from a simple RPC scenario:

```
{  
   "ref":{  
      "channelId":"3"
   },
   "data":{  
      "state":{  
         "state":"READY"
      },
      "target":"localhost:29788",
      "trace":{  
         "numEventsLogged":"4",
         "creationTimestamp":"2018-10-15T13:45:51.809444890Z",
         "events":[  
            {  
               "description":"Channel created",
               "severity":"CT_INFO",
               "timestamp":"2018-10-15T13:45:51.809445533Z"
            },
            {  
               "description":"Channel state change to IDLE",
               "severity":"CT_INFO",
               "timestamp":"2018-10-15T13:45:51.809445011Z"
            },
            {  
               "description":"Channel state change to CONNECTING",
               "severity":"CT_INFO",
               "timestamp":"2018-10-15T13:45:51.809444963Z"
            },
            {  
               "description":"Channel state change to READY",
               "severity":"CT_INFO",
               "timestamp":"2018-10-15T13:45:51.810445024Z"
            }
         ]
      },
      "callsStarted":"1",
      "callsSucceeded":"1",
      "lastCallStartedTimestamp":"2018-10-15T13:45:51.809444632Z"
   },
   "subchannelRef":[  
      {  
         "subchannelId":"4"
      },
      {  
         "subchannelId":"5"
      }
   ]
}
```

### CODE REVIEW

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE                                                                      FILE SIZE
 ++++++++++++++ GROWING                                                        ++++++++++++++
  +0.1%    +664 [None]                                                         +6.95Ki  +0.1%
  +4.3%    +336 src/core/ext/filters/client_channel/subchannel.cc                 +336  +4.3%
      [NEW]    +329 set_subchannel_connectivity_state_locked                          +329  [NEW]
      +6.4%     +19 [Unmapped]                                                         +19  +6.4%
  +0.9%    +320 src/core/ext/filters/client_channel/client_channel.cc             +320  +0.9%
      +134%    +292 set_channel_connectivity_state_locked                             +292  +134%
      +1.2%     +16 cc_init_channel_elem                                               +16  +1.2%
      [NEW]     +12 grpc_client_channel_set_channelz_node                              +12  [NEW]
  +1.0%     +16 src/core/ext/filters/client_channel/client_channel_channelz.cc     +16  +1.0%
       +13%     +16 grpc_core::channelz::ClientChannelNode::ClientChannelNode          +16   +13%

  +0.1% +1.30Ki TOTAL                                                          +7.61Ki  +0.1%


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
Objective-C binary sizes
*****************STATIC******************
  New size                      Old size
 1,966,767      Total (>)      1,965,922

 No significant differences in binary sizes

***************FRAMEWORKS****************
  New size                      Old size
10,909,695      Total (>)     10,908,996

 No significant differences in binary sizes


```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

File: src/core/ext/filters/client_channel/client_channel.cc [hunk 1]
```diff
@@ -699,6 +724,7 @@ static grpc_error* cc_init_channel_elem(grpc_channel_element* elem,
   // Record enable_retries.
   arg = grpc_channel_args_find(args->channel_args, GRPC_ARG_ENABLE_RETRIES);
   chand->enable_retries = grpc_channel_arg_get_bool(arg, true);
+  chand->channelz_channel = nullptr;
```
Comment (markdroth) [inline src/core/ext/filters/client_channel/client_channel.cc]: No need for this; `chand` is initialized to all zero.

Comment (ncteisen) [reply src/core/ext/filters/client_channel/client_channel.cc]: I think @soheilhy is working to change this invariant in the name of performance.

Mind if I keep to make his life easier?

Comment (soheilhy) [reply src/core/ext/filters/client_channel/client_channel.cc]: Yes, thank you for initialization it! I'm actually preparing the PR at the moment. :-)

Comment (markdroth) [reply src/core/ext/filters/client_channel/client_channel.cc]: Okay, that's fine.

@soheilhy, I will be very interested in reviewing that PR.  Thanks for doing it!

Comment (soheilhy) [reply src/core/ext/filters/client_channel/client_channel.cc]: Absolutely, I'd really appreciate your time and feedback :-) ETA is (hopefully) tomorrow morning.

**Pull request merged**